### PR TITLE
[Clang] Start moving X86Builtins.def to X86Builtins.td

### PR DESCRIPTION
**Pull request by philnik777**
This starts moving `X86Builtins.def` to be a tablegen file. It's quite large, so I think it'd be good to move things in multiple steps to avoid a bunch of merge conflicts due to the amount of time this takes to complete.

### CODE REVIEW

**Comment by RKSimon** [comment]
@philnik777 Is this part of a larger scheme covering other targets?

**Comment by philnik777** [comment]
> @philnik777 Is this part of a larger scheme covering other targets?

Ultimately all the builtins should be in TableGen files instead of the `.def` files, but that's going to take a while. Probably years, since it's not exactly high priority.

**Comment by philnik777** [comment]
gentle ping~

**Review by AaronBallman — COMMENTED** [review]
I think this is a reasonable direction to go, so the changes notionally LGTM. And I didn't spot any issues with the conversion either, but I'd appreciate a second set of eyes more familiar with x86 builtins to double-check.

File: clang/include/clang/Basic/BuiltinsX86.td [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    def pmulhuw128 : X86Builtin<"_Vector<8, short>(_Vector<8, short>, _Vector<8, short>)">;
+    def vec_ext_v2di : X86Builtin<"long long int(_Vector<2, long long int>, _Constant int)">;
+    def vec_ext_v4si : X86Builtin<"int(_Vector<4, int>, _Constant int)">;
+    def vec_ext_v4sf : X86Builtin<"float(_Vector<4, float>, _Constant int)">;
+    def vec_ext_v8hi : X86Builtin<"short(_Vector<8, short>, _Constant int)">;
+    def vec_set_v8hi : X86Builtin<"_Vector<8, short>(_Vector<8, short>, short, _Constant int)">;
+  }
+
+  foreach Op = ["addsub", "hadd", "hsub"] in {
+    let Features = "sse3" in {
```
Comment (RKSimon) [inline clang/include/clang/Basic/BuiltinsX86.td:109]: why is the sse3 on the inside of the loop here? match ssse3 below?

**Comment by RKSimon** [comment]
No objections to this patch, I do want to see other targets moved over as well though, instead of x86 ending up being treated as a special/weird case.

I am preparing to convert a lot of x86 builtins to be constexpr (#30794) - but as long as this PR won't interfere I see no big reason not to get it done right away.

**Comment by philnik777** [comment]
> No objections to this patch, I do want to see other targets moved over as well though, instead of x86 ending up being treated as a special/weird case.

That is the plan. It just takes time to move everything. FWIW Hexagon and RISCV already use TableGen files as well.

> I am preparing to convert a lot of x86 builtins to be constexpr (#30794) - but as long as this PR won't interfere I see no big reason not to get it done right away.

It shouldn't. You just have to add the `Constexpr` attribute instead of whatever you had to do in the `.def` files.

**Comment by philnik777** [comment]
@AaronBallman @RKSimon does this look good to you?

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Review by RKSimon — APPROVED** [review]

**Pull request merged**